### PR TITLE
Feat: Support for explicitly typed attributes in Authentication (#308)

### PR DESCRIPTION
**Pull request by logineimer**
Adds support for explicitly typed attributes in Authentication like @p1729 suggested.

### CODE REVIEW

File: security/src/main/java/io/micronaut/security/authentication/Authentication.java [hunk 1]
```diff
@@ -40,4 +41,18 @@
...
+
+    /**
+     * Retrieves an explicitly typed attribute value.
+     *
+     * @param name the name of the attribute
+     * @param clazz the expected class of the attribute
+     * @param <T> the expected type of the attribute
+     * @return the typed attribute or null if it doesn't exist or the type is not correct
+     */
+    default <T> Optional<T> getAttribute(String name, @NonNull Class<T> clazz) {
```
Comment (sdelamo) [inline security/src/main/java/io/micronaut/security/authentication/Authentication.java:53]: Could you annotate the return type with `@NonNull`

Could you annotate the parameter `name` with `@NonNull`

**Review by sdelamo — CHANGES_REQUESTED** [review]

**Comment by sdelamo** [comment]
Thanks for the contribution. 

However, I am not sure if we should add this method. I think the recommended way do conversion is to use the [ConversionService](https://docs.micronaut.io/latest/api/index.html).

**Comment by sdelamo** [comment]
@jameskleeh thoughts ?

**Comment by jameskleeh** [comment]
The issue with this is that authentication objects are created outside the context and we would need a conversion service instance to do the conversion. There is ConversionService.SHARED we could use but it wouldn't include any converters added via beans, so less than ideal.

**Comment by sdelamo** [comment]
I am going to close it. My recommendation is for the user to inject `ConversionService` in their singleton or bean where they are querying the authentication. I don't want to add a method to `Authentication`which does not represent best practice.

**Pull request closed without merging**